### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Limits): add classes `ReflectsFiniteLimits` and friends

### PR DESCRIPTION
**Pull request by dagurtomas**
This PR adds classes `ReflectsFiniteLimits`, `ReflectsFiniteProducts` and their duals, and relates them to existing classes.

It also cleans up the file `CategoryTheory.Limits.Preserves.Finite` a bit and removes some porting notes that are no longer relevant.

---
This is used in #12870 to prove that to check the coherent sheaf condition on a finitary extensive preregular category, it is enough to check after postcomposition with a finite-limit-reflecting functor.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Limits/Preserves/Finite.lean [hunk 1]
```diff
@@ -156,8 +219,6 @@ noncomputable instance (priority := 120) PreservesColimits.preservesFiniteColimi
   PreservesColimitsOfSize.preservesFiniteColimits F
 #align category_theory.limits.preserves_colimits.preserves_finite_colimits CategoryTheory.Limits.PreservesColimits.preservesFiniteColimits
 
--- Porting note: is this unnecessary given the instance
--- `PreservesColimitsOfSize0.preservesFiniteColimits`?
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Limits/Preserves/Finite.lean:160]: This porting note was based on the misunderstanding that `PreservesColimitsOfSize0.preservesFiniteColimits` and `preservesFiniteColimitsOfPreservesFiniteColimitsOfSize` have anything to do with each other.

**Comment by dagurtomas** [comment]
!bench

File: Mathlib/CategoryTheory/Limits/Preserves/Finite.lean [hunk 2]
```diff
@@ -4,7 +4,7 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Andrew Yang
 -/
 import Mathlib.CategoryTheory.Limits.Preserves.Basic
-import Mathlib.CategoryTheory.FinCategory.AsType
+import Mathlib.CategoryTheory.Limits.Shapes.FiniteProducts
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Limits/Preserves/Finite.lean:7]: Without this import change, this file doesn't know about the classes `HasFiniteLimits` and `HasFiniteProducts` etc.

**Comment by dagurtomas** [comment]
!bench

File: Mathlib/CategoryTheory/Limits/Preserves/Finite.lean [hunk 3]
```diff
@@ -90,13 +88,13 @@ def preservesFiniteLimitsOfPreservesFiniteLimitsOfSize (F : C ⥤ D)
 #align category_theory.limits.preserves_finite_limits_of_preserves_finite_limits_of_size CategoryTheory.Limits.preservesFiniteLimitsOfPreservesFiniteLimitsOfSize
 
 noncomputable instance idPreservesFiniteLimits : PreservesFiniteLimits (𝟭 C) :=
-  ⟨fun _ _ _ => by infer_instance⟩
+  inferInstance -- should this still be stated explicitly an instance?
 #align category_theory.limits.id_preserves_finite_limits CategoryTheory.Limits.idPreservesFiniteLimits
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Limits/Preserves/Finite.lean:92]: I would suggest you try to remove it!

Comment (dagurtomas) [reply Mathlib/CategoryTheory/Limits/Preserves/Finite.lean]: Ok, let's see what happens!

**Comment by joelriou** [comment]
Thanks!

bors d+

**Comment by dagurtomas** [comment]
bors merge

**Pull request closed without merging**